indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskManager.java

```diff
@@ -0,0 +1,721 @@
+
+    if (disabled.compareAndSet(false, true)) {
+      changeHistory.addChangeRequest(new WorkerHistoryItem.Metadata(true));
+    }
+  }
+
+  private static class TaskDetails
+  {
+    private final Task task;
+    private final long startTime;
+    private TaskStatus status;
+    private TaskLocation location;
+
+    public TaskDetails(Task task)
+    {
+      this.task = task;
+      this.startTime = System.currentTimeMillis();
+      this.status = TaskStatus.running(task.getId());
+      this.location = TaskLocation.unknown();
+    }
+  }
+
+  private interface Notice
+  {
+    String getTaskId();
+
+    void handle() throws Exception;
+  }
+
+  private class RunNotice implements Notice
+  {
+    private final Task task;
+
+    public RunNotice(Task task)
+    {
+      this.task = task;
+    }
+
+    @Override
+    public String getTaskId()
+    {
+      return task.getId();
+    }
+
+    @Override
+    public void handle() throws Exception
+    {
+      TaskAnnouncement announcement = null;
+      synchronized (lock) {
+        if (runningTasks.containsKey(task.getId()) || completedTasks.containsKey(task.getId())) {
+          log.warn(
+              "Got run notice for task [%s] that I am already running or completed...",
+              task.getId()
+          );
+
+          taskStarted(task.getId());
+          return;
+        }
+
+        final ListenableFuture<TaskStatus> future = taskRunner.run(task);
+        addRunningTask(task, future);
+
+        announcement = TaskAnnouncement.create(
+            task,
+            TaskStatus.running(task.getId()),
+            TaskLocation.unknown()
+        );
+
+        changeHistory.addChangeRequest(new WorkerHistoryItem.TaskUpdate(announcement));
+
+        cleanupAssignedTask(task);
+        log.info("Task[%s] started.", task.getId());
+      }
+
+      taskAnnouncementChanged(announcement);
+      taskStarted(task.getId());
+    }
+  }
+
+  private class StatusNotice implements Notice
+  {
+    private final Task task;
+    private final TaskStatus status;
+
+    public StatusNotice(Task task, TaskStatus status)
+    {
+      this.task = task;
+      this.status = status;
+    }
+
+    @Override
+    public String getTaskId()
+    {
+      return task.getId();
+    }
+
+    @Override
+    public void handle() throws Exception
+    {
+      synchronized (lock) {
+        final TaskDetails details = runningTasks.get(task.getId());
+
+        if (details == null) {
+          log.warn("Got status notice for task [%s] that isn't running...", task.getId());
+          return;
+        }
+
+        if (!status.isComplete()) {
+          log.warn(
+              "WTF?! Got status notice for task [%s] that isn't complete (status = [%s])...",
+              task.getId(),
+              status.getStatusCode()
+          );
+          return;
+        }
+
+        details.status = status.withDuration(System.currentTimeMillis() - details.startTime);
```

Comment:
Hey @himanshug, I'm currently looking at the codes around task runtime to make it available while tasks are running (it is available only after the task is done for now). I noticed that [the task runtime is first computed in peon](https://github.com/apache/druid/blob/master/indexing-service/src/main/java/org/apache/druid/indexing/overlord/SingleTaskBackgroundRunner.java#L444), but overwritten here. Do you remember why it's overwriting?

Reply:
I guess it's possible that the duration could be missing if the peon process died unexpectedly?

Reply:
sorry, noticed it now, I think it makes sense as duration computed by peon wouldn't take into account things such as jvm startup, other setup things, shutdown etc... measurement here is much more accurate and representative of peon's/task's lifetime.

Reply:
Hmm interesting. What do you think about having finer-granular states such as `ASSIGNED` (assigned to a middleManager or an indexer) and `STARTED` (same as current `RUNNING`), so that we can measure task runtime and peon startup time separately?

Reply:
nothing wrong with measuring more granular timestamps if there is use for it. also, if you just need the time taken by task, more specifically the time taken by `Task.run()` then maybe publish that as a  metric from `ThreadingTaskRunner` and maybe that is good enough.